### PR TITLE
Introduce the bufr_query library from NOAA-EMC

### PR DESCRIPTION
**Pull request by srherbener**
This PR adds in a new package.py script for the new bufr_query library from NOAA-EMC. This is being used by JEDI and other applications.

This partially addresses jcsda/spack-stack/issues/1182

### CODE REVIEW

File: var/spack/repos/builtin/packages/bufr-query/package.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    depends_on("py-pybind11", type=("build"), when="+python")
+
+    # CMake configuration
+    def cmake_args(self):
+        args = [self.define_from_variant("BUILD_PYTHON_BINDINGS", "python")]
+
+        # provide path to netcdf-c include files
+        nc_include_dir = Executable("nc-config")("--includedir", output=str).strip()
+        args.append("-DCMAKE_C_FLAGS=-I" + nc_include_dir)
+        args.append("-DCMAKE_CXX_FLAGS=-I" + nc_include_dir)
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/bufr-query/package.py:52]: Why is that needed? If bufr-query is a `cmake` package, it should run `find_package(netcdf)` itself, or?

File: var/spack/repos/builtin/packages/bufr-query/package.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    depends_on("eckit@1.24.4:", type=("build", "run"))
+    depends_on("eigen@3:", type=("build", "run"))
+    depends_on("gsl-lite", type=("build", "run"))
+    depends_on("netcdf-c", type=("build", "run"))
+    depends_on("netcdf-cxx4", type=("build", "run"))
+    depends_on("bufr", type=("build", "run"))
+
+    # Optional dependencies
+    variant("python", default=True, description="Enable Python interface")
+    depends_on("python@3:", type=("build", "run"), when="+python")
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/bufr-query/package.py]: I think that the correct syntax for Python is something like `extends(python)` and then use the PythonBuilder to build and install the Python interface. See https://github.com/spack/spack/pull/45504/files for how this is done. I haven't implemented a package this way myself yet, but I understand that this is how it is supposed to be.

At the minimum, please remove `@3:` because the minimum version that spack supports is `3.7`.

**Review by climbfuji — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 1 above)
Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py:52]: This is what I found in other formulas to fix the problem where the netcdf.h file cannot be found. Here is the find_package in the bufr_query CMakeLists.txt.

https://github.com/NOAA-EMC/bufr-query/blob/a78d4c79d72aef3a7046386090ee5c3c51554718/CMakeLists.txt#L30

Im guessing that asking for only the CXX component of the netcdf library is the issue. However, it seems that asking for only the CXX component should realize that you have to also use the C component and the netcdf find package process would bring that in for you. It must be attempting to do that but the include path is not getting set - once that is done it finds the libraries just fine.

I can verify if this truly is the issue.

What's the best way to proceed?
- I could replace this code with a patch that adds the C component to the bufr_query netcdf find_package call
- OR We could try to get the fix into bufr_query and remove this code
- OR leave it as is

Thanks!

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 2 above)
Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: I grabbed this from another package script too. The example you give looks a lot nicer. I can take a look at that. Thanks!

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 1 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/bufr-query/package.py:52]: I suggest leaving as is and filing a bug in the bufr_query repo to fix this. We had to do this (explicitly add the C dependency even if needed for a Fortran dependency) for other components in the past, too (OpenMP in FMS for macOS).

Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py:52]: Adding the C component in the netcdf find_package call in the bufr-query CMakeLists.txt did not help - still got the cannot file "netcdf.h" include file. But we can leave as is for now (as noted above).

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 2 above)
Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: I switched the code over to use the extend("python") mechanism (as is done in the example PR you noted).

**Review by climbfuji — APPROVED** [review]
Great that you got this to work so quickly. I haven't tested this myself, but I am happy to approve if this works for you.

**Comment by srherbener** [comment]
Tested this and verified that the bufr_query install is working. This is good to go now.

I solved the PYTHONPATH issue with a patch, and at the same time I've submitted a PR to bufr-query with the changes in the patch.

**Review by climbfuji — APPROVED** [review]
Approving again. Looks good to me

**Comment by climbfuji** [comment]
The commit history is a bit too long for merging this as a regular merge commit (which we usually do for merges into the spack submodule for easier cherry-picking for upstream spack develop). I am going to merge this as a squash merge and we'll have to do something like `git checkout jcsda/spack-stack-dev -- var/spack/repos/builtin/packages/bufr_query` to simply grab the whole directory and push it upstream to spack develop.

**Comment by srherbener** [comment]
I have created a PR in the authoritative spack repo with these changes: https://github.com/spack/spack/pull/45920
